### ISSUE TITLE
New product copies are immediately sent to Facebook

### ISSUE THREAD

**Issue by grig-23** [woocommerce/facebook-for-woocommerce#1803]
## 🔎 Isolate the bug

- [x] I have confirmed this occurs in the most recent version of WordPress, WooCommerce, and Facebook for WooCommerce.
- [x] I have confirmed this occurs when only WooCommerce and Facebook for WooCommerce are active and when using a default WordPress or WooCommerce theme.

## ✍️ Describe the bug
When a synced product is copied to a new draft, a copy is immediately synced to Facebook. Quite often, this workflow is used to create completely new and different products. It doesn't mean that duplicate products are also needed on the Facebook side though.

##🚶‍♀️ Steps to reproduce

1. Copy a synced product to a new draft.
2. Change all the details about the product including SKUs.
3. Don't publish the product.
4. There will be a duplicate on Facebook side.

## ✔️ Expected behavior
Ideally, products should only be synced to Facebook after they're published on WooCommerce side so that users get a chance to disable Facebook sync before publishing.

**Comment by budzanowski** [comment]
The syncing of the product in the draft state should not happen, as long as the product is in the draft state the synchronization is disabled. This is indicated by

**Comment by tuijatuulia** [comment]
Hi, Originally my problem was not "Ideally, products should only be synced to Facebook after they're published on WooCommerce side so that users get a chance to disable Facebook sync before publishing."

The problem is, that I translate all multivendor site products, and when I publish the product, it also goes to Facebook **right away**. There is no possibility to not-sync it first. In happy-day scenario nice, but in practise, you don't want all products to be included, like test products, gift cards, user groups, products without price (services, custom-made), and digital products, and those that FB does not allow, At least some of these all should have - at least test products in every release round.

It would be better to 
a) schedule the sync to happen at certain times to allow time to "please do not sync this" or 
b) mark the products manually that you want to be synced in FB (so by default NOT synced is on)

Additionally, I have very annoying trouble with staging site, which syncs to FB and I cannot stop it (deactivated plugin etc

**Comment by mitchellkrogza** [comment]
> The syncing of the product in the draft state should not happen, as long as the product is in the draft state the synchronization is disabled. This is indicated by the `Do not sync` in the `Facebook Sync` column in all products view:
> ![image](https://user-images.githubusercontent.com/17271089/111168097-07fab880-85a2-11eb-805f-7ac3d5f9e124.png)
> 
> This will also be visible in the Facebook metabox on the right-hand side of the product page ( below product tags ):
> ![image](https://user-images.githubusercontent.com/17271089/111168348-48f2cd00-85a2-11eb-9c1e-8cfa80cfc223.png)
> 
> Could you please provide us the system status report: https://docs.woocommerce.com/document/understanding-the-woocommerce-system-status-report/#section-1
> 
> Please also share the screenshots of the previously mentioned admin interface areas and how is the product set up. End the exact procedure that you follow that causes the duplications. This will greatly help us understand and fix the problem.
> 
> Thank you for your help!

I am the one who originally raised this issue on the plugin support forum and it was posted here by @grig-23   

It's not practical to have a pre-made product that is set to not sync. Why? 

My products have +- 70 variations which means when I do create a new product I will have to go to 70 variations and set them to sync. Not feasible nor practical in any way. Nor can I set the whole product and all variations to sync as 2 variations are virtual and cannot be synced to FB shops. 

Woocommerce has a "duplicate" product option which when used creates a new "draft" product but this plugin syncs and makes copies causing a total mess of the Facebook store. 

My daily workflow, as I add new product daily is to just disable this plugin create my new product and then re-enable it. It works doing this but is a very backwards way of working. 

A sync schedule "might" be a feature but the plugin should really adhere to what Woocommerce is doing and that is creating a new draft which is unpublished and should not be synced.

**Comment by beckily** [comment]
I am facng the same issue and would be delighted to find a quick and clean solution to it.

**Comment by mitchellkrogza** [comment]
> I am facing the same issue and would be delighted to find a quick and clean solution to it.

My daily workflow is
- Deactivate the plugin
- Add your new product(s)
- When you are done activate the plugin and all new stuff syncs instantly

**Comment by solstudioim** [comment]
Another user reported the same issue [here](https://wordpress.org/support/topic/draft-products-keep-showing/).

**Comment by zeinnicholas** [comment]
I can also confirm the same issue. My clients have their products generated by their ERP via RestAPI, and set to Pending Review. But the products are being synced directly to both Facebook and Instagram Shops. Then the client clicks on the link to buy the product on the website and the connection times out because of too many redirects (obviously, since the product is not yet published on the website). This has negatively affected our Analytics and user engagement. Moreover, since the products are not published yet, and don't have images, they get synchronized with the default placeholder image. Not even our custom image is synced.

This has been happening for a while now. At this point you should have already found the issue and fixed it.

@budzanowski what is the ETA for fixing this issue?

**Comment by theabhig** [comment]
Also reported in 4003413-zen.

**Comment by danielbitzer** [comment]
I've tested this and can reproduce this issue. Steps to reproduce:

1. Create or choose a **variable** product with variations set to all variations to "sync and show in catalog"
2. In the **WooCommerce > Products** list view click the **Duplicate** link for the variable product
3. Don't save or publish the duplicate
3. Wait for a while and eventually the duplicate product will also appear in the Facebook Catalogue view despite having a draft status

<img width="980" alt="Screen Shot 2021-05-20 at 5 48 09 pm" src="https://user-images.githubusercontent.com/1892137/118944386-ac7aeb80-b993-11eb-975d-a703d0424536.png">

**Comment by zeinnicholas** [comment]
You don't need to change the variation setting to "sync and show in catalog" for this to happen. All you need to do is to create a product or variation and not define anything in said setting, which is what is happening to me. The products and variations are being created with the bare minimum and required information, which means custom meta like the FB sync setting is not being inserted. One could argue that then we should include that information when creating products via RESTAPI, but it shouldn't be the default value for this particular setting. The default value should be to not sync, and include conditions to sync once the product is published and visible in the catalog, making sure that variations don't override the parent product.

**Comment by budzanowski** [comment]
@danielbitzer I was looking through the sync code while checking the validator class and something caught my attention. I was wondering what is the status of the duplicated variations ( not the variable product but the inner variations ). It turns out that although the parent post is 'draft' the variations are 'publish' - this is the WC core behaviour so I doubt that we could influence this. But we can check the parent when we check the variation. What do you think: https://github.com/woocommerce/facebook-for-woocommerce/pull/1971 ?

I hope that this actually is all what would we need to change to fix this.

**Comment by danielbitzer** [comment]
@budzanowski I think the start of the issue is that the `\WC_Facebookcommerce_Integration::on_product_save` method assumes that the product being saved is `published`. But the product can have any status at this point. 

It's pretty hard to see what's supposed to be happening with all this code though. I'm thinking a possibly good, and safe, solution would be to enhance the logic for the determining product `visibility` on this line:

https://github.com/woocommerce/facebook-for-woocommerce/blob/9b1a1dcb696fe4e1926b9f196b0375705fad4c41/includes/fbproduct.php#L576

and for some reason it's copy/pasted to this line: (not sure why)

https://github.com/woocommerce/facebook-for-woocommerce/blob/9b1a1dcb696fe4e1926b9f196b0375705fad4c41/includes/fbproduct.php#L594

Here we could check that the product and it's parent all have the `published` status. Let me know what you think.

**Comment by budzanowski** [comment]
@danielbitzer I think that visibility would not be enough. The product would still be synced as a copy. The situation would be improved with the product not being exposed to clients but still, it would be synced as a `Product ( Copy )` which from what I see the merchants would want to avoid. Until they update the name and change all other details that they need the product should no be synchronized. In the end, they may want to even abandon the copy.

Or I am not seeing this correctly and visibility hidden would actually stop the sync?

Hard to tell what is the best way to do this. The first step would be to understand what are all the flows that are triggering the sync.
You are right this is a 🍝.

**Comment by zeinnicholas** [comment]
> @danielbitzer I think that visibility would not be enough. The product would still be synced as a copy. The situation would be improved with the product not being exposed to clients but still, it would be synced as a `Product ( Copy )` which from what I see the merchants would want to avoid. Until they update the name and change all other details that they need the product should no be synchronized. In the end, they may want to even abandon the copy.
> 
> Or I am not seeing this correctly and visibility hidden would actually stop the sync?
> 
> Hard to tell what is the best way to do this. The first step would be to understand what are all the flows that are triggering the sync.
> You are right this is a 🍝.

@budzanowski I think the proper flow that would cover all needs would be the following:

1. Products should only be automatically synced if:
    - Product Status is **Published**
    - Catalog Visibility is **Visible**
    - Product is **_not_ Password Protected**
    - Product is **_not_ Private**
    - Product Status is **_not_ Pending Review**
    - Product Status is **_not_ Draft**
2. Variations (child product) should not sync unless the parent Product of type "Variable" meets item 1 above.
3. _**IF**_ the Shop Manager wants, they should be able to manually set individual products or variations to sync by manually setting the "Facebook sync" option. (But honestly I don't think it should be possible to set it when the product status is not "published"). 

This should cover all cases, and products in Facebook should at least be "Deactivated" when they don't meet the parameters above.

**Comment by jessLundie** [comment]
Hi team! I was also able to replicate this with a regular draft product created through normal methods (not *Duplicate*). These are the steps I followed:

1. **Products** > Add Product
2. Variable Product
3. Add an attribute
4. Create variations for all attributes automatically
5. Add prices to each variant, but *do not* adjust the Facebook status in any way
6. Save product as draft

When you sync products to Facebook, you get all the variants synced, but without pricing information:
![https://d.pr/i/IFSywU](https://d.pr/i/IFSywU+)
Full Image: https://d.pr/i/IFSywU

On the WooCommerce side, the product shows as Do Not Sync:
![https://d.pr/i/seb6zD](https://d.pr/i/seb6zD+)
Full Image: https://d.pr/i/seb6zD

Customer in 4006886-zen reports that this is also affecting products created through 3rd party "pushes" - I've asked for more details, but I'm assuming these are products created via the REST API that come in in draft form. Instant-sync to Facebook.

**Comment by johndcoy** [comment]
I could not replicate this issue using the REST API cURL method for `Draft` or `Pending` statuses. Both Simple and Variable products did not publish to Facebook as expected. However, we have had a report of the "push generator" within Printful has this problem, as well as the report from @zeinnicholas 

@zeinnicholas could you provide insight into the API method used to create products with the `pending` status?

**Comment by zeinnicholas** [comment]
> I could not replicate this issue using the REST API cURL method for `Draft` or `Pending` statuses. Both Simple and Variable products did not publish to Facebook as expected. However, we have had a report of the "push generator" within Printful has this problem, as well as the report from @zeinnicholas
> 
> @zeinnicholas could you provide insight into the API method used to create products with the `pending` status?

I don't think this is user error... But anyways...
They are using the API v2. Here's a complete example:

**POST /wc/v2/products**
```
{
    "name": "Flat Pietra Violeta ",
    "type": "variable",
    "status": "pending",
    "regular_price": "520",
    "stock_quantity": "18",
    "description":"Flat Pietrar
        Pelica Violeta Lezard r
        100% Couro e solado em couro",
    "in_stock": "true",
    "sku": "0490178",
    "attributes": [
        {
            "id": "1",
            "name": "TAMANHO",
            "position": "0",
            "visible": "true",
            "variation": "true",
            "options": [
                "33",
                "34",
                "35",
                "36",
                "37",
                "38",
                "39",
                "40"
            ]
        },
        {
            "id": "2",
            "name": "COR",
            "position": "1",
            "visible": "true",
            "variation": "false",
            "options": [
                "Violeta "
            ]
        },
        {
            "id": "3",
            "name": "SALTO - Nu00DAMERO",
            "position": "2",
            "visible": "true",
            "variation": "false",
            "options": [
                "1"
            ]
        },
        {
            "id": "5",
            "name": "SALTO - ALTURA",
            "position": "1",
            "visible": "true",
            "variation": "false",
            "options": [
                "Baixo"
            ]
        },
        {
            "id": "6",
            "name": "SALTO - TIPO",
            "position": "1",
            "visible": "true",
            "variation": "false",
            "options": [
                "Outros"
            ]
        }
    ],
    "categories": [
        {
            "id": 1012
        }
    ]
}
```

**POST /wc/v2/products/605208/variations**
```
{
    "parent_id": "605208",
    "regular_price": "520",
    "manage_stock": "true",
    "stock_quantity": "0",
    "description":"Flat Pietrar
        Pelica Violeta Lezard r
        100% Couro e solado em couro",
    "in_stock": "false",
    "sku": "049017833",
    "weight": "1",
    "dimensions": {
        "length": "28",
        "width": "18",
        "height": "9"
    },
    "attributes": [
        {
            "id": "1",
            "option": "33"
        }
    ]
}
```

**POST /wc/v2/products/605208/variations**
```
{
    "parent_id": "605208",
    "regular_price": "520",
    "manage_stock": "true",
    "stock_quantity": "0",
    "description":"Flat Pietrar
        Pelica Violeta Lezard r
        100% Couro e solado em couro",
    "in_stock": "false",
    "sku": "049017834",
    "weight": "1",
    "dimensions": {
        "length": "28",
        "width": "18",
        "height": "9"
    },
    "attributes": [
        {
            "id": "1",
            "option": "34"
        }
    ]
}
```

**POST /wc/v2/products/605208/variations**
```
{
    "parent_id": "605208",
    "regular_price": "520",
    "manage_stock": "true",
    "stock_quantity": "2",
    "description":"Flat Pietrar
        Pelica Violeta Lezard r
        100% Couro e solado em couro",
    "in_stock": "true",
    "sku": "049017835",
    "weight": "1",
    "dimensions": {
        "length": "28",
        "width": "18",
        "height": "9"
    },
    "attributes": [
        {
            "id": "1",
            "option": "35"
        }
    ]
}
```

**POST /wc/v2/products/605208/variations**
```
{
    "parent_id": "605208",
    "regular_price": "520",
    "manage_stock": "true",
    "stock_quantity": "5",
    "description":"Flat Pietrar
        Pelica Violeta Lezard r
        100% Couro e solado em couro",
    "in_stock": "true",
    "sku": "049017836",
    "weight": "1",
    "dimensions": {
        "length": "28",
        "width": "18",
        "height": "9"
    },
    "attributes": [
        {
            "id": "1",
            "option": "36"
        }
    ]
}
```

**POST /wc/v2/products/605208/variations**
```
{
    "parent_id": "605208",
    "regular_price": "520",
    "manage_stock": "true",
    "stock_quantity": "6",
    "description":"Flat Pietrar
        Pelica Violeta Lezard r
        100% Couro e solado em couro",
    "in_stock": "true",
    "sku": "049017837",
    "weight": "1",
    "dimensions": {
        "length": "28",
        "width": "18",
        "height": "9"
    },
    "attributes": [
        {
            "id": "1",
            "option": "37"
        }
    ]
}
```

**POST /wc/v2/products/605208/variations**
```
{
    "parent_id": "605208",
    "regular_price": "520",
    "manage_stock": "true",
    "stock_quantity": "3",
    "description":"Flat Pietrar
        Pelica Violeta Lezard r
        100% Couro e solado em couro",
    "in_stock": "true",
    "sku": "049017838",
    "weight": "1",
    "dimensions": {
        "length": "28",
        "width": "18",
        "height": "9"
    },
    "attributes": [
        {
            "id": "1",
            "option": "38"
        }
    ]
}
```

**POST /wc/v2/products/605208/variations**
```
{
    "parent_id": "605208",
    "regular_price": "520",
    "manage_stock": "true",
    "stock_quantity": "2",
    "description":"Flat Pietrar
        Pelica Violeta Lezard r
        100% Couro e solado em couro",
    "in_stock": "true",
    "sku": "049017839",
    "weight": "1",
    "dimensions": {
        "length": "28",
        "width": "18",
        "height": "9"
    },
    "attributes": [
        {
            "id": "1",
            "option": "39"
        }
    ]
}
```

**POST /wc/v2/products/605208/variations**
```
{
    "parent_id": "605208",
    "regular_price": "520",
    "manage_stock": "true",
    "stock_quantity": "0",
    "description":"Flat Pietrar
        Pelica Violeta Lezard r
        100% Couro e solado em couro",
    "in_stock": "false",
    "sku": "049017840",
    "weight": "1",
    "dimensions": {
        "length": "28",
        "width": "18",
        "height": "9"
    },
    "attributes": [
        {
            "id": "1",
            "option": "40"
        }
    ]
}
```

**POST /wc/v2/products/605208/variations/batch**
```
{"update":[{"id":605196,"regular_price":"520","sale_price":"","stock_quantity":"5","in_stock":"true"}]}
```

**POST /wc/v2/products/605208/variations/batch**
```
{"update":[{"id":605223,"regular_price":"520","sale_price":"","stock_quantity":"5","in_stock":"true"}]}
```

**Comment by theabhig** [comment]
Similar issue reported in 4050474-zen - the user had a new test product get sent to Facebook, which wasn't a duplicate.

**Comment by atikullamerchant** [comment]
Has anyone been able to find a workaround?

**Comment by anant1811** [comment]
Also reported in 4056004-zen

**Comment by zeinnicholas** [comment]
So... It's been more than 3 months now. Is the issue fixed with 2.6.0? My customers are complaining!

@danielbitzer?

**Comment by haszari** [comment]
Facebook for WooCommerce 2.6.0 included an improved validator class, which checks whether a product should be included in sync. See #1933 for details.

I've retested the steps from @jessLundie above and confirmed that this issue is fixed by the new validator class. In 2.5.1 my draft child variations sync when they shouldn't; in 2.6 and newer they don't sync when not published. 🎉 

Here's a brief rundown of the testing steps I was using. Note I was only able to trigger the issue when doing a full sync via `Sync products` admin button.

##### Create a draft variable product, with some variations with prices
- Click `Products > Add New`.
- Give it a distinctive product name so you can find it in Facebook catalog if it syncs.
- Scroll down to `Product data` box, select `Variable product` from product type dropdown.
- Click `Attributes`. Select an attribute, click `Add`.
- Add attribute values (`Select terms` box).
- Ensure `Used for variations` is checked.
- Click `Save attributes`.
- Click `Variations`. 
- Select `Create variations from all attributes`, click `Go`, and then `OK` the warning modal.
- Give each new variation a price.
- Scroll up and click `Save Draft`.

##### Sync all products manually
- Go to `Marketing > Facebook`.
- Click `Product sync` tab.
- Click `Sync all products`.

### Closing!
I'm closing this issue, as the behaviour has been improved in Facebook for WooCommerce 2.6.0.

If anyone is seeing similar issues on their store (with 2.6.0), please open an new issue with steps to reproduce the bug. Thank you 🙇

**Comment by mitchellkrogza** [comment]
> Facebook for WooCommerce 2.6.0 included an improved validator class, which checks whether a product should be included in sync. See #1933 for details.
> 
> I've retested the steps from @jessLundie above and confirmed that this issue is fixed by the new validator class. In 2.5.1 my draft child variations sync when they shouldn't; in 2.6 and newer they don't sync when not published. 🎉
> 
> Here's a brief rundown of the testing steps I was using. Note I was only able to trigger the issue when doing a full sync via `Sync products` admin button.
> 
> ##### Create a draft variable product, with some variations with prices
> * Click `Products > Add New`.
> * Give it a distinctive product name so you can find it in Facebook catalog if it syncs.
> * Scroll down to `Product data` box, select `Variable product` from product type dropdown.
> * Click `Attributes`. Select an attribute, click `Add`.
> * Add attribute values (`Select terms` box).
> * Ensure `Used for variations` is checked.
> * Click `Save attributes`.
> * Click `Variations`.
> * Select `Create variations from all attributes`, click `Go`, and then `OK` the warning modal.
> * Give each new variation a price.
> * Scroll up and click `Save Draft`.
> 
> ##### Sync all products manually
> * Go to `Marketing > Facebook`.
> * Click `Product sync` tab.
> * Click `Sync all products`.
> 
> ### Closing!
> I'm closing this issue, as the behaviour has been improved in Facebook for WooCommerce 2.6.0.
> 
> If anyone is seeing similar issues on their store (with 2.6.0), please open an new issue with steps to reproduce the bug. Thank you 🙇

You needed one more step in your test ... To take the product you created click "duplicate" which creates a new draft and see if the new validator prevents an instant sync of the duplicated product which is what's happening with many users. Duplicate creates a new draft of an existing product but it is immediately synced to FB with the words (Copy) in the title.

**Comment by haszari** [comment]
> You needed one more step in your test ... To take the product you created click "duplicate" which creates a new draft and see if the new validator prevents an instant sync of the duplicated product which is what's happening with many users. Duplicate creates a new draft of an existing product but it is immediately synced to FB with the words (Copy) in the title.

Thanks for following up @mitchellkrogza - I appreciate your attention to detail!

I just retested this:

- Regular or variable products (with variations), marked as `Do not sync`
- `Duplicate`ed each of these products, and also `Save draft` in product screen

I didn't see any of them sync to Facebook, so 🤞 this is fixed. Please open a new issue with steps if you can reproduce a similar issue!

**Comment by mitchellkrogza** [comment]
Awesome @haszari I shall test the new version and confirm.

**Comment by mitchellkrogza** [comment]
:clap: @haszari I can confirm this is resolved in 2.6.0
- I duplicated a product this morning
- State of the draft product became Do Not Sync
- left it for a while before publishing while watching my logs to make sure nothing was happening in the background
- published the new product
- state set to Sync and Show and within 3 minutes fetched by Facebook :clap:

**Comment by haszari** [comment]
Props go to @danielbitzer for [tidying up how products are validated before sync](https://github.com/woocommerce/facebook-for-woocommerce/pull/1933) 🏅
